File: infra/lib/metric-names.ts
/**
 * Single source of truth for the Phase-6 metric contract that the Monitoring
 * stack (and its tests) reference when building CloudWatch dashboards/alarms.
 *
 * These are the *dotted* meter names exactly as emitted by Micrometer in the
 * serving API (see
 * `service/topsales-api/.../metrics/MetricNames.java`). The CloudWatch registry
 * publishes them dotted, under the `topsales-api` namespace.
 *
 * NOTE ON RENDERING: the local/`prometheus` registry renders the same meters
 * with dots → underscores and appends `_total` to counters
 * (e.g. `topsales.read.total` → `topsales_read_total`). Dashboards that target
 * Prometheus must use the underscore form; CloudWatch uses the dotted form
 * below. Keep both stacks pointing at THIS object so a rename can't drift.
 */
export const METRIC_NAMES = {
  /** Counter: read-path responses, tagged `status` (fresh|stale|degraded|pending) and `mode`. */
  READ_TOTAL: 'topsales.read.total',

  /** Counter: insight generations that fell back to the deterministic template. */
  INSIGHT_FALLBACK: 'topsales.insight.fallback.total',

  /** Gauge: seconds since the newest serving-row `as_of` (global) — forecast staleness. */
  FORECAST_FRESHNESS_SECONDS: 'topsales.forecast.freshness.seconds',

  /** Counter: serving-read faults swallowed by the fail-soft read path (a dropped degradation rung). */
  PROVIDER_FAULT: 'topsales.forecast.provider.faults.total',

  /** Actuator/Micrometer built-in: RED latency/throughput/errors for HTTP requests. */
  HTTP_SERVER_REQUESTS: 'http.server.requests',
} as const;

export type MetricName = (typeof METRIC_NAMES)[keyof typeof METRIC_NAMES];

File: service/topsales-api/src/main/java/com/topsales/api/metrics/MetricNames.java
package com.topsales.api.metrics;

/**
 * Canonical names for the custom Phase-6 Micrometer meters (scraped at {@code /actuator/prometheus}).
 * RED latency/throughput/errors come free from actuator's {@code http.server.requests}; these are the
 * domain-specific ML-quality and freshness signals layered on top.
 *
 * <p>Held as constants in one place so the emitting sites (read path, insight layer, freshness gauge)
 * and any dashboards/tests reference the exact same string — a typo'd meter name is otherwise a silent
 * gap that only shows up as a missing Prometheus series.
 */
public final class MetricNames {

    private MetricNames() {}

    /** Counter: read-path responses, tagged {@code status} (fresh|stale|degraded|pending) and {@code mode}. */
    public static final String READ_TOTAL = "topsales.read.total";

    /** Counter: insight generations that fell back to the deterministic template (emitted by the insight layer). */
    public static final String INSIGHT_FALLBACK = "topsales.insight.fallback.total";

    /** Gauge: seconds since the newest serving-row {@code as_of} (global) — forecast staleness. */
    public static final String FORECAST_FRESHNESS_SECONDS = "topsales.forecast.freshness.seconds";

    /** Counter: serving-read faults swallowed by the fail-soft read path (a dropped degradation rung). */
    public static final String PROVIDER_FAULT = "topsales.forecast.provider.faults.total";
}
